File: occasum-core/src/main/java/pw/stamina/occasum/node/AbstractParentedPropertyNode.java
package pw.stamina.occasum.node;

import pw.stamina.occasum.PropertyHandle;
import pw.stamina.occasum.node.factory.PropertyNodeFactory;

import java.util.Objects;

abstract class AbstractParentedPropertyNode extends AbstractPropertyNode {
    private final PropertyNode parent;

    AbstractParentedPropertyNode(PropertyNodeFactory factory,
                                 PropertyHandle handle,
                                 PropertyNode parent) {
        super(factory, handle);
        Objects.requireNonNull(parent, "parent");
        this.parent = parent;
    }

    @Override
    public final PropertyNode getParent() {
        return parent;
    }

    @Override
    public final boolean hasParent() {
        return true;
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/node/AbstractParentedPropertyNode.kt
package pw.stamina.occasum.node

import pw.stamina.occasum.PropertyHandle
import pw.stamina.occasum.node.factory.PropertyNodeFactory

internal abstract class AbstractParentedPropertyNode(
        factory: PropertyNodeFactory,
        handle: PropertyHandle,
        override val parent: PropertyNode
) : AbstractPropertyNode(factory, handle)
